iceoryx2-pal/posix/src/scandir.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use crate::posix::types::*;
+use crate::posix::{closedir, free, malloc, opendir, readdir_r};
+
+pub unsafe fn scandir_impl(
+    path: *const c_char,
+    namelist: *mut *mut *mut crate::posix::types::dirent,
+) -> int {
+    let dirfd = opendir(path);
+    if dirfd.is_null() {
+        return -1;
+    }
+
+    *namelist = core::ptr::null_mut::<*mut dirent>();
+    let mut entries = vec![];
+    const DIRENT_SIZE: usize = core::mem::size_of::<dirent>();
+
+    let cleanup = |entries: &Vec<*mut void>, namelist: *mut *mut dirent| {
+        for entry in entries {
+            free((*entry).cast());
+        }
+
+        if !namelist.is_null() {
+            free(namelist.cast());
+        }
+    };
+
+    loop {
+        let dirent_ptr = malloc(DIRENT_SIZE);
+        let result_ptr: *mut *mut dirent = malloc(core::mem::size_of::<*mut dirent>()).cast();
+
+        if readdir_r(dirfd, dirent_ptr.cast(), result_ptr) != 0 {
+            free(result_ptr.cast());
+            free(dirent_ptr);
+            cleanup(&entries, *namelist);
```

Comment:
If you agree with the cleanup changes, this also needs to be changed
```suggestion
            cleanup(&mut entries, namelist);
```